test/python/test_whoami_api.py

```diff
@@ -0,0 +1,38 @@
+from __future__ import absolute_import
+
+import unittest
+
+import openapi_client
+
+from . import api_config
+
+WHOAMI_FIELDS = [
+    "client_ip",
+    "user_agent",
+    "account",
+    "username",
+    "token_issued_at",
+]
+
+class TestWhoamiApi(api_config.ConfiguredTest):
+    """WhoamiApi unit test stubs"""
+
+    def setUp(self):
+        self.api = openapi_client.api.whoami_api.WhoamiApi(self.client)
+
+    def test_who_am_i(self):
```

Comment:
I think we should try to test several different things for this endpoint, and probably the others.
- Verify that a good request gets a 200 response
- Verify that a good request gets the proper response body (currently we just verify that we get a body with a non-empty value for each attribute, which is probably good enough for now)
- Verify that a bad request gets a 401 response
- Verify that a bad request doesn't return any of the '201 attributes', or if it returns an error, we should check for that

Reply:
I do in some cases check that a bad request returns an error code. However isn't that to an extent out of scope for integration tests? I thought we were just supposed to be testing to make sure that the client works properly with the API. That seems like something that unit tests for the API would handle, but I could be way off about that.

Reply:
So that would be in-scope, if these are integration tests. I'll over-explain why I think that:

- Unit tests focus on a _specific piece of code_, verifying input and output at that stage. In our case, unit tests would be making sure that a function within the API returns the piece of information we expect. So, for example, within the `authn_api` for Ruby, we would want to verify that `login(account, opts = {})` returns the expected output when given specific input for `account` and `opts`. These almost always should work offline.

- Integration Tests, or "black box" testing focuses on passing input in to the product, as you would in a production environment, and verifying the output is what you would receive in production. These typically require a production style environment.

In this case, if we want to verify that the generated spec performs as expected, we would need integration tests that verify the input detailed in the spec also produces the output detailed in the spec, as it would in a production environment.